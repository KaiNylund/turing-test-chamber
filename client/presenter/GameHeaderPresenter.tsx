import React, { useState } from "react";
import { getTitleText } from "../utils/GameStateUtils";
import GameHeaderView from "../view/GameHeaderView";
import GameInfoView from "../view/GameInfoView";

function GameHeaderPresenter(props: {roleType: "loading" | "p1" | "p2" | "guesser",
                                     isGuessing: boolean,
                                     guessOutcome: "undecided" | "correct" | "incorrect"}) {
  const [darkMode, setDarkMode] = useState(localStorage.getItem("isDarkmode") === "true");
  const [showInfo, setShowInfo] = useState(localStorage.getItem("returningUser") === "true");

  localStorage.setItem("returningUser", "true");

  function toggleDarkMode() {
    let toggleMode = !darkMode;
    setDarkMode(toggleMode)
    localStorage.setItem("isDarkmode", toggleMode.toString());
  }

  function toggleGameInfo() {
    setShowInfo(!showInfo);
  }

  const titleText = getTitleText(props.isGuessing,
                                 props.guessOutcome,
                                 props.roleType);

    if (darkMode) {
      document.getElementById("root")?.classList.add("dark");
    } else {
      document.getElementById("root")?.classList.remove("dark");
    }

  return (
    <div>
      <GameHeaderView titleText={titleText}
                    darkMode={darkMode}
                    toggleDarkMode={toggleDarkMode}
                    toggleGameInfo={toggleGameInfo}/>
      <GameInfoView hidden={showInfo}
                    toggleInfo={toggleGameInfo}/>
    </div>
  );
}

export default GameHeaderPresenter;