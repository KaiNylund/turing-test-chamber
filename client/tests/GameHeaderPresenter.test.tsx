import React from 'react';
import renderer from 'react-test-renderer';
import GameHeaderPresenter from '../presenter/GameHeaderPresenter';
import { JSDOM } from "jsdom";

const dom = new JSDOM()
global.document = dom.window.document

function storageMock() {
  let storage: any = {"returningUser": "false"};

  return {
    clear: () => {
      storage = {};
    },
    setItem: function(key: string | number, value: string) {
      storage[key] = value || '';
    },
    getItem: function(key: string) {
      return key in storage ? storage[key] : null;
    },
    removeItem: function(key: string | number) {
      delete storage[key];
    },
    get length() {
      return Object.keys(storage).length;
    },
    key: function(i: number) {
      const keys = Object.keys(storage);
      return keys[i] || null;
    }
  };
}

global.localStorage = storageMock();

const component = renderer.create(
  <GameHeaderPresenter
    roleType={'guesser'}
    isGuessing={false}
    guessOutcome={'undecided'}/>
);
let tree: any = component.toJSON();
let headerChildren = tree.children[0].children;
let gameInfoPage = tree.children[1].children

// Test game info toggle
it('gameInfo close works', async () => {
  let closeImgComp = gameInfoPage.find((e: { type: string; }) => e.type === "header").children[1];
  expect(tree).toMatchSnapshot();

  await renderer.act(async () => {
    closeImgComp.props.onMouseDown();
  });

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});


// Test dark mode toggle
it('dark mode toggle works', async () => {
  let darkModeComp = headerChildren.find((e: { type: string; }) => e.type === "div");
  expect(tree).toMatchSnapshot();

  await renderer.act(async () => {
    darkModeComp.props.onMouseDown();
  });

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

// Test game info toggle
it('dark mode toggle works', async () => {
  let imgToggleComp = headerChildren.find((e: { type: string; }) => e.type === "img");
  expect(tree).toMatchSnapshot();

  await renderer.act(async () => {
    imgToggleComp.props.onMouseDown();
  });

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});