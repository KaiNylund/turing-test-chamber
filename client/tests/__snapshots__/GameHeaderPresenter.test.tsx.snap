// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dark mode toggle works 1`] = `
<div>
  <header>
    <img
      onMouseDown={[Function]}
      src=""
      width={23}
    />
    <h1>
      You are the guesser
    </h1>
    <div
      className="battery light"
      onMouseDown={[Function]}
    />
  </header>
</div>
`;

exports[`dark mode toggle works 2`] = `
<div>
  <header>
    <img
      onMouseDown={[Function]}
      src=""
      width={23}
    />
    <h1>
      You are the guesser
    </h1>
    <div
      className="battery dark"
      onMouseDown={[Function]}
    />
  </header>
</div>
`;

exports[`dark mode toggle works 3`] = `
<div>
  <header>
    <img
      onMouseDown={[Function]}
      src=""
      width={23}
    />
    <h1>
      You are the guesser
    </h1>
    <div
      className="battery dark"
      onMouseDown={[Function]}
    />
  </header>
</div>
`;

exports[`dark mode toggle works 4`] = `
<div>
  <header>
    <img
      onMouseDown={[Function]}
      src=""
      width={23}
    />
    <h1>
      You are the guesser
    </h1>
    <div
      className="battery dark"
      onMouseDown={[Function]}
    />
  </header>
</div>
`;

exports[`gameInfo close works 1`] = `
<div>
  <header>
    <img
      onMouseDown={[Function]}
      src=""
      width={23}
    />
    <h1>
      You are the guesser
    </h1>
    <div
      className="battery light"
      onMouseDown={[Function]}
    />
  </header>
  <div
    id="game-info"
  >
    <header>
      <h1>
        How to play
      </h1>
      <img
        onMouseDown={[Function]}
        src=""
        width={20}
      />
    </header>
    <p>
      This game is based on the Turing test, where a player (our "guesser") has to figure out which of two other players (our "tricker" and GPT-3 model) is a computer using only their responses to written questions.
    </p>
    <section>
      <h2>
        Guesser
      </h2>
      <p>
        If you are the guesser, then it's your job to figure out whether p1 or p2 is the computer. Start by sending a text to both players using the input bar and see if you can tell from their responses. After you've exchanged a few texts and you're ready to make your guess, click the "Guess!" button and select the player you think is the computer.
      </p>
    </section>
    <section>
      <h2>
        Tricker
      </h2>
      <p>
        If you are trying to trick the guesser, then it's your job to make them think you are the computer. GPT-3 isn't great for these kinds of tasks, so we've given you the advantage of seeing the first part of its response. It's up to you if you want to imitate what GPT-3 is saying, or come up with your own answers.
      </p>
    </section>
    <p>
      Once the guesser has made their guess, both players will be able to see the results and statistics about their game. Good luck!
    </p>
  </div>
</div>
`;

exports[`gameInfo close works 2`] = `
<div>
  <header>
    <img
      onMouseDown={[Function]}
      src=""
      width={23}
    />
    <h1>
      You are the guesser
    </h1>
    <div
      className="battery light"
      onMouseDown={[Function]}
    />
  </header>
</div>
`;
